### PR TITLE
Implement AWS IMDSv2 support

### PR DESCRIPTION
**Pull request by nmeyerhans**
### Relevant Issue (if applicable)
#1445 

### Details

AWS IMDSv2 is a session oriented method for retrieving instance metadata,
including IAM credentials, in Amazon EC2.  It is enabled by default in
non-enforcing mode in AWS (meaning it retains backwards compatibility with
existing IMDSv1 clients), but can be switched to enforcing mode, in which
clients are required to return API tokens with requests.

With this change, we implement support for IMDSv2 and enable it by default when
IAM roles are our source for authentication credentials.  In the event that
s3fs is running in cloud environment offering an IMDSv1-compatible API, we
support graceful fallback to that mode.  It can also be selected explicitly via
the imdsv1only mount option.

More details on IMDSv2 are available at
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
and
https://aws.amazon.com/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/

### CODE REVIEW

**Comment by nmeyerhans** [comment]
I've revised this PR to avoid using the C++11 `std::to_string` function when converting from an integer value. This should address the earlier test failure.

File: src/curl.cpp [hunk 1]
```diff
@@ -2701,6 +2756,23 @@ int S3fsCurl::GetIAMCredentials()
         }
         url = std::string(S3fsCurl::IAM_cred_url) + env;
     }else{
+		if(S3fsCurl::IAM_api_version > 1){
```
Comment (gaul) [inline src/curl.cpp:2759]: Fix strange indentation (tabs vs. 4-spaces?)

File: src/s3fs_help.cpp [hunk 2]
```diff
@@ -313,6 +313,13 @@ static const char help_string[] =
     "      \"auto\", s3fs will automatically use the IAM role names that are set\n"
     "      to an instance. If you specify this option without any argument, it\n"
     "      is the same as that you have specified the \"auto\".\n"
+	"\n"
+	"   imdsv1only (default is to use IMDSv2)\n"
```
Comment (gaul) [inline src/s3fs_help.cpp:317]: Please add to `doc/man/s3fs.1` as well.

File: src/curl.h [hunk 3]
```diff
@@ -348,6 +356,7 @@ class S3fsCurl
         static void InitUserAgent();
         static bool SetRequesterPays(bool flag) { bool old_flag = S3fsCurl::requester_pays; S3fsCurl::requester_pays = flag; return old_flag; }
         static bool IsRequesterPays() { return S3fsCurl::requester_pays; }
+		static bool SetIMDSVersion(int version);
```
Comment (gaul) [inline src/curl.h:359]: Fix indentation.

File: src/curl.cpp [hunk 4]
```diff
@@ -1656,6 +1668,13 @@ bool S3fsCurl::ParseIAMCredentialResponse(const char* response, iamcredmap_t& ke
     return true;
 }
 
+bool S3fsCurl::SetIAMv2APIToken(const char* response)
+{
+    S3FS_PRN_INFO3("Setting AWS IMDSv2 API token to %s", response);
+    S3fsCurl::IAMv2_api_token = std::string(response);
+	return true;
```
Comment (gaul) [inline src/curl.cpp:1675]: Indentation.

**Review by gaul — CHANGES_REQUESTED** [review]
Please address small comments.  Overall this seems fine.

**Comment by gaul** [comment]
Thank you for your contribution @nmeyerhans!

**Pull request merged**